text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,772 @@
+
+For example, how do Ember apps express that they depend on a third-party library? The [app.import](https://ember-cli.com/user-guide/#javascript-assets) API. This should be ECMA standard `import`.
+
+Another way to state the problem is that apps and addons all _push_ whatever code they want into the final built app. Whereas ES modules can _pull_ each other into the build as needed.
+
+### Play nice with NPM Conventions
+
+The ECMA module spec by itself doesn’t try to define a module resolution algorithm. But the overwhelmingly most popular convention is the [node_modules resolution algorithm](https://nodejs.org/api/all.html#modules_all_together).
+
+Ember addons do respect node_module resolution for build-time code, but they do not respect it for runtime code. This is an unhelpful distinction.
+
+### Verbose, Static Javascript as a Compiler Target
+
+Ember’s strong conventions mean that many kinds of dependencies can be inferred (including _statically_ inferred) without requiring the developer to laboriously manage them. This is a good thing and I believe the current fad in the wider Javascript ecosystem for making developers hand-write verbose static imports for everything confuses the benefits of having static analysis (which is good) with the benefits of hand-managing those static imports (which is unnecessary cognitive load when you have clear conventions and a compiler).
+
+This design is about compiling today’s idiomatic Ember code into more “vanilla” patterns that leverage ES modules, node_modules resolution, and spec-compliant static and dynamic `import` to express the structure of an Ember application in a much more “vanilla Javascript” way.
+
+This compile step lets us separate the authoring format (which isn’t changing in any significant way in this RFC) from the packaging format (which can be more verbose and static than we would want in an authoring format).
+
+# Detailed design
+
+## Definitions
+
+**package**: every addon and app is a package. Usually synonymous with “NPM package”, but we also include in-repo packages. The most important fact about a package is that it’s often the boundary around code that comes from a particular author, team, or organization, so coordination across packages is a more sensitive design problem than coordination within apps.
+
+**app**: a package used at the root of a project.
+
+**addon**: a package not used at the root of a project. Will be an **allowed dependency** of either an **app** or an **addon**.
+
+**allowed dependency**: For **addons**, the **allowed dependencies** are the `dependencies` and `peerDependencies` in `package.json` plus any in-repo addons. For **apps**, the **allowed dependencies** are the `dependencies`, `peerDependencies`, and `devDependencies` in `package.json` plus any in-repo addons.
```

Comment:
I recognize that we want to keep `devDependencies` as an **allowed dependency** for **apps**, since this is how it's always been conventionally for Ember **apps** and we want to reduce churn as much as possible.

Out of curiosity I would still like to understand, why it ever was this way. Is it just a side-effect of the fact that **apps** are always the root of any given project and thus will always have access to `devDependencies`?

I would argue that, while still officially supporting it, we should discourage adding runtime dependencies for **apps** to `devDependencies` instead of `dependendencies`, for the following reasons:

- It is confusing for Ember newbies.
- It is inconsistent.
- When I turn an **app** into an engine (an **addon**) or extract certain parts of an **app** into an **addon** (or engine), `devDependencies` suddenly turn into `dependencies`, just because the code was moved.

Instead, I would recommend that we encourage the same mental model for **apps** and **addons**. Everything that needs to be available at runtime / should influence the production build, has to be a `dependency`. Everything else, like testing or development support, linters, etc. are `devDependencies`.

What is your opinion regarding this?

If you agree, we could have an opt-in flag to enforce the same semantics for **apps** and **addons**.

Reply:
I strongly agree with this but would go a step further: if at all possible, I would very much like to see us get to a world where devDependencies and dependencies mean what they sound like: the latter is genuinely runtime-only, the former genuinely development- and build-time only.

It’s possible there are reasons that cannot be the case, in which case I’d love clarification there, because my understanding of the three-phase build pipeline for Embroider suggests that precisely that distinction should be possible. But I very well may be missing something!

Reply:
Yes, for addons the distinction gets quite strong under this RFC. And it will reduce the number of packages needed by apps (count how many copies of ember-cli-htmlbars and ember-cli-babel you have).

For apps, I agree with:

>  we should discourage adding runtime dependencies for apps to devDependencies instead of dependencies

But most apps don't have any runtime dependencies. If you do `ember b`, and then `rm -rf node_modules`, and then serve `dist` to a browser, your app still runs. There are no runtime dependencies. Your addons may be used at runtime, but the actual addon NPM packages are only needed during the build.

The exception is Fastboot. Fastboot does imply having some runtime dependencies. Using `dependencies` for this would simplify and improve fastboot, which today tries to generate a whole separate package.json.